### PR TITLE
Feature: show upload date in the network content window

### PR DESCRIPTION
**Pull request by perezdidac**
## Motivation / Problem

Upload date is present in every content in BaNaNaS but not used in OpenTTD since it's not sent to the client. https://github.com/OpenTTD/bananas-server/pull/43 changes it so it gets sent.

## Description

This code change makes OpenTTD show the upload date for a given NewGRF in the online content window. In addition, a column in the content list has been added to sort by upload date as well.

image

## Limitations

See https://github.com/OpenTTD/bananas-server/pull/43 for full details, but in particular I've focused on making sure this doesn't break older versions. Here is what I have successfully tested:

1. Bananas returning the new field, OpenTTD not knowing about it -> OpenTTD ignores it as it has read the latest field it knows about it (tags) and moves on to the next NewGRF. No memory leak or anything out of normal.
2. Bananas returning the new field, OpenTTD knowing about it -> OpenTTD reads it successfully.
3. ~~Bananas not returning the new field, OpenTTD expecting it -> OpenTTD tries to read it but does not actually read it. I used CanReadFromPacket() to verify and make it safe in case we have to rollback Bananas and newer OpenTTD clients still want the value.~~ finally didn't end up including this, see conversation below.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by LordAro** [comment]
Re: date formatting stuff. There already exists a setting for default date formats in savegame names, see `date_format_in_default_names` & `STR_JUST_DATE_*` constants. Perhaps the setting could be repurposed/renamed to be used more generally, and use the same string formatting mechanisms that are used for that?

**Comment by perezdidac** [comment]
> Re: date formatting stuff. There already exists a setting for default date formats in savegame names, see `date_format_in_default_names` & `STR_JUST_DATE_*` constants. Perhaps the setting could be repurposed/renamed to be used more generally, and use the same string formatting mechanisms that are used for that?

Sounds good, I'll do that, but first I need to implement a function to parse the date as it comes and convert it to a Date object. I will work on that first and once that works I'll use date_format_in_default_names to decide how to display it and potentially rename it to something more generic as you suggest. Thanks!

**Comment by perezdidac** [comment]
I finally managed to make it work through passing a UNIX epoch timestamp.
It works great and backwards compatibility in case BaNaNaS rolls back or the other way around, old OpenTTD versions are not impacted.
I am still looking at how to format using the suggestions @LordAro mentioned above but I couldn't figure it out, so I left it as 2021-06-23 (ymd) for now. See screenshot.

File: src/network/core/packet.cpp [hunk 1]
```diff
@@ -180,6 +180,15 @@ bool Packet::CanReadFromPacket(uint bytes_to_read)
 	return true;
 }
 
+/**
+ * Returns the number of bytes left to read to reach the end of the packet
+ * @return The amount of bytes left to be read
+ */
+int Packet::GetBytesLeftInPacket()
```
Comment (TrueBrain) [inline src/network/core/packet.cpp]: Personally, I really wouldn't do this. It mostly adds complexity to the code that will never be used.

If we decide to change the protocol, that is an irreversible change. We upgrade the bananas-server first, merge this PR after, and that is kinda final for the protocol.

If we ever revert this functionality, it cannot revert the protocol, as we have clients that now depend on it. Patchpacks might already have copied it etc. So even if we revert, these bytes will remain reserved in the protocol till the end of time.

Say we revert and also remove the bytes from the protocol, and some time later someone adds another field to the protocol, older clients (and patchpacks that copied the functionality) will still read that new field as upload-date, which can only result in other problems. This is why not having a version field is so annoying btw :p

In other words: protocol changes are final and always only extending (never removing), and clients don't need code "in case the protocol is reverted". That is an impossible situation.

Again, at least, this is my opinion :) feel free to proof me wrong :D

Comment (nielsmh) [reply src/network/core/packet.cpp]: Agree on this. If you're going to extend the protocol, do it in an extensible way. For example use a tagged format like RIFF, which is also used for the savegame format (although only chunk-level is tagged there). Then the extended protocol would start with a "magic" tag, followed by number of bytes in the extended protocol part, followed by each add-on data element tagged.
But at that point you may as well replace the entire protocol.

Comment (TrueBrain) [reply src/network/core/packet.cpp]: That is not what I said, so just to make sure there is no confusion about what I said:

I am perfectly fine with this PR, except for the part it keeps backwards compatibility with an old bananas-server. That is a scenario that simply is never going to happen, so I see no benefit of having code for that :) Just assuming this new field is always there, is more than sufficient in my opinion.
This mostly as I can fully imagine patchpacks releasing a version which include this PR, at which point reverting is impossible. So better already assume it is impossible, and build the code accordingly.

That is all I was saying :D Rewriting the full protocol should happen some day, but is, as far as I am concerned, out of scope for this PR.

Comment (perezdidac) [reply src/network/core/packet.cpp]: Thanks both. @nielsmh we discussed the extensibility of the data passing in https://github.com/OpenTTD/bananas-server/pull/43, and I think everyone would agree that we need to make a proper rewrite implementing this over JSON/proto or another more structured data messaging framework that also lets us pass versioning, etc.

But I can't do this now, it's too hard of a task and I mainly want to get this PR in so we can have this feature first. I honestly don't mind volunteering to get the long-term approach done, but that will require a proper design proposal and many approvals from the dev community.

Coming back to the topic: thanks @TrueBrain for clarifying. If you **really** believe I shouldn't worry about the possibility of having to revert BaNaNaS then I am fine removing it. Nicer, cleaner, and forward-looking, yeah!

**Comment by perezdidac** [comment]
Hi all! I am still looking for more devs to look into it. Based on conversations with @TrueBrain they seem to agree with this but we need other devs to comment on this. Here is the PR in bananas-server that will get sent first if this moves forward: https://github.com/OpenTTD/bananas-server/pull/43

**Comment by LordAro** [comment]
What's the advantage to sorting by upload date?

**Comment by perezdidac** [comment]
> What's the advantage to sorting by upload date?

That's the question I was expecting :D

I can see a use case when as a user I want to browse random NewGRFs and know if they are fairly new or years old. In my case particularly, let's say I wanted a nice industry set to spice the game up and I get a few options, it is nice to know if it has been recently updated or the last release was 10 years ago. I am not saying old NewGRFs are implemented poorly or are bad quality or anything like that, I just think it's a feature some players will benefit from. At least to me, it's important to know I am using NewGRFs that are newer and fairly maintained.

https://bananas.openttd.org/package/newgrf seems to sort by default as well.

**Comment by LordAro** [comment]
That's all true, but I don't feel there's any benefit to actually sorting by upload date - displaying it in the general info is perfectly fine though

**Comment by perezdidac** [comment]
> That's all true, but I don't feel there's any benefit to actually sorting by upload date - displaying it in the general info is perfectly fine though

Oh I remember why I added that to this PR, I did it given that https://github.com/OpenTTD/OpenTTD/issues/8903 was opened and deduped against this one.

**Comment by perezdidac** [comment]
It's been a while. @LordAro what do you think about this feature at this point? should we move forward without the ability to sort or want to include it? I'm inclined to as the player would benefit from the ability to find latest releases.

**Comment by 2TallTyler** [comment]
I think this is still worth considering, but it needs a rebase. 🙂

**Comment by TrueBrain** [comment]
I think we should just do this. Not the sorting, as that will give annoying problems with people rating newer content as better than older, etc. But having the information in there, should lovely.

I wouldn't call it `upload date`, but rather `last updated`. Although internally it is ofc the last upload date, but using `last updated` give more the illusion that there is one item, with multiple versions. Instead of considering every version its own entity.

And yes, this needs a chain of changes in other services too, which all have open PRs, but just untouched for .. 2 years? now (sorryyyyyyyyyy). But in general, I think this is just a nice addition.

**Comment by TrueBrain** [comment]
Although I am still a fan of this change, I think we can also reasonable deduce that this PR has stalled to the point it is not moving forward. And no developer has picked it up.

So I am sorry, and I do appreciate the effort, but I am going to close up this PR.

Again, sorry for the long delay, and I do appreciate the effort you put into this. If you ever feel like picking this up again, please start here, and once we are okay with this, we will do the backend work.
